lib/Driver/ToolChains.cpp

```diff
@@ -526,6 +401,185 @@ ToolChain::constructInvocation(const CompileJobAction &job,
+    Arguments.push_back("-supplementary-output-file-map");
+    Arguments.push_back(getTemporaryFilePath("supplementaryOutputs", ""));
+    FilelistInfos.push_back({Arguments.back(), types::TY_INVALID,
+                             FilelistInfo::WhichFiles::SupplementaryOutput});
+  } else {
+    addFrontendSupplementaryOutputArguments(Arguments);
+  }
+}
+
+void ToolChain::JobContext::addFrontendInputArguments(
+    const bool mayHavePrimaryInputs, const bool useFileList,
+    const bool usePrimaryFileList, const bool filterByType,
+    ArgStringList &arguments) const {
+  llvm::StringSet<> primaries;
+
+  if (mayHavePrimaryInputs)
+    for (const Action *A : InputActions) {
+      const auto *IA = cast<InputAction>(A);
+      const llvm::opt::Arg &InArg = IA->getInputArg();
+      // For an index data job,
+      // only make the index-file-path ones primary
+      if (ToolChain::canCompileInputArgumentBePrimary(Output, InArg))
+        primaries.insert(InArg.getValue());
+    }
+  for (auto inputPair : getTopLevelInputFiles()) {
+    if (filterByType && !types::isPartOfSwiftCompilation(inputPair.first))
+      continue;
+    const char *inputName = inputPair.second->getValue();
+    if (primaries.count(inputName)) {
+      if (!usePrimaryFileList) {
+        arguments.push_back("-primary-file");
+        arguments.push_back(inputName);
+      }
+    } else {
+      if (!useFileList) {
+        arguments.push_back(inputName);
+      }
+    }
+  }
+}
+
+bool ToolChain::canCompileInputArgumentBePrimary(const CommandOutput &Output,
+                                                 const llvm::opt::Arg &A) {
+  return Output.getPrimaryOutputType() != types::TY_IndexData ||
+         A.getOption().getID() == options::OPT_index_file_path;
+}
+
+void ToolChain::JobContext::addFrontendSupplementaryOutputArguments(
+    ArgStringList &arguments) const {
+  // FIXME: get these and other argument strings from the same place for both
+  // driver and frontend
+  addOutputsOfType(arguments, Output, Args, types::ID::TY_SwiftModuleFile,
+                   "-emit-module-path");
+
+  addModuleDocPathArgIfNecessary(Output, Args, arguments);
+
+  addSerializeDiagnosticsPathArgIfNecessary(Output, Args, arguments);
+
+  if (addOutputsOfType(arguments, Output, Args, types::ID::TY_ObjCHeader,
+                       "-emit-objc-header-path")) {
+    assert(OI.CompilerMode == OutputInfo::Mode::SingleCompile &&
+           "The Swift tool should only emit an Obj-C header in single compile"
+           "mode!");
+  }
+
+  addOutputsOfType(arguments, Output, Args, types::TY_Dependencies,
+                   "-emit-dependencies-path");
+  addOutputsOfType(arguments, Output, Args, types::TY_SwiftDeps,
+                   "-emit-reference-dependencies-path");
+  addOutputsOfType(arguments, Output, Args, types::TY_ModuleTrace,
+                   "-emit-loaded-module-trace-path");
+  addOutputsOfType(arguments, Output, Args, types::TY_TBD, "-emit-tbd-path");
```

Comment:
…or `-emit-tbd-path`?

Reply:
The TBDPath appears to be determined by `chooseTBDPath` which is only called for a `CompileJobAction`: 
      if (isa<CompileJobAction>(JA))
        chooseDependenciesOutputPaths(C, OI, OutputMap, workingDirectory, Buf,
                                  Output.get());
So it seemed unnecessary.

Reply:
This one is needed for SingleCompile but currently not supported in StandardCompile, so I guess it's okay.